### ISSUE TITLE
Chapel 1.30.0 (llvm 14.0.6) issues with CUDA 12.x (deprecated) texture template?

### ISSUE THREAD

**Issue by hpcpony** [chapel-lang/chapel#22085]
```
[hpcpony@host2 Chapel]$ chpl --version
warning: The prototype GPU support implies --no-checks. This may impact debuggability. To suppress this warning, compile with --no-checks explicitly
chpl version 1.30.0
  built with LLVM version 14.0.6
Copyright 2020-2023 Hewlett Packard Enterprise Development LP
Copyright 2004-2019 Cray Inc.
(See LICENSE file for more details)

[hpcpony@host2 Chapel]$ cat gpu.chpl

writeln("here.gpus ", here.gpus);

[hpcpony@host2 Chapel]$ chpl gpu.chpl
warning: The prototype GPU support implies --no-checks. This may impact debuggability. To suppress this warning, compile with --no-checks explicitly
In file included from <built-in>:1:
In file included from /public/opt/chapel-1.30.0_host2/third-party/llvm/install/linux64-x86_64/lib/clang/14.0.6/include/__clang_cuda_runtime_wrapper.h:365:
/public/opt/chapel-1.30.0_host2/third-party/llvm/install/linux64-x86_64/lib/clang/14.0.6/include/__clang_cuda_texture_intrinsics.h:696:13: error: no template named 'texture'
            texture<__DataT, __TexT, cudaReadModeNormalizedFloat> __handle,
            ^
/public/opt/chapel-1.30.0_host2/third-party/llvm/install/linux64-x86_64/lib/clang/14.0.6/include/__clang_cuda_texture_intrinsics.h:709:13: error: no template named 'texture'
            texture<__DataT, __TexT, cudaReadModeElementType> __handle,
            ^
2 errors generated when compiling for sm_80.
error: error running clang during code generation
```
I found the following but I'm not sure if it's relevant:

https://forums.developer.nvidia.com/t/cuda-12-0-still-support-for-texture-reference-support-for-pascal-architecture-warp-synchronous-programming/237284

My Chapel build on a CUDA 11.7 machine seems to work fine (i.e. compiles and finds gpus).

**Comment by e-kayrakli** [comment]
@hpcpony -- thanks for the bug report!

We need to look a bit into it, but I suspect we are limited by clang as it doesn't support CUDA versions beyond 11.5: https://llvm.org/docs/CompileCudaWithLLVM.html#prerequisites

We have been suppressing warnings generated by it so far, because it defaults to dropping to CUDA 11.5 and it works fine for all of Chapel's uses of CUDA as long as the major version is 11.x. However, that does not hold with CUDA 12.

Short term actions for us are:
- Be more judicious when suppressing warnings
  - This is probably something we can do at build time, instead of waiting for clang's warning to pop up and turn it into an error.
- Document the limitation in our [GPU technote](https://chapel-lang.org/docs/main/technotes/gpu.html)

In the meantime, is dropping back to CUDA 11.7 problematic for you beyond the annoyance of installing that on your system? If the clang documentation is correct there, I am not entirely sure what we can do besides waiting for clang to support the new CUDA version.

**Comment by hpcpony** [comment]
I'm not an expert (nor fan) of LLVM but from poking around I'd agree that LLVM is the issue.  At a minimum the chapel build should probably check the CUDA version and fault if it's 12.x.

As for what to do...  That kinda depends on another question I have about the "best practices" for GPU installs.  What I don't understand is how much of the chapel build relies on the CUDA that is specified at build time vs. how much (CUDA, arch, ...?) can be deferred and specified at application compile time.  This isn't clear from the CPU technote.  It would sure make life easier if I could build one GPU version of chapel and then mount it on the various GPU hosts and specify the CUDA/arch at application time.  The alternative is that I have to build for each host/CUDA/arch version.

I did try using the CUDA 12.1 build but pointed it at application compile time to a (user) installed CUDA 11.8 and it seemed to work but that was only the simple "hello world" above.

The bottom line is I can (user level) install a different version of CUDA, but I would not be able to do a (system level) downgrade from 12.x to 11.x.

Beyond that I'm open to suggestions. ;^)

**Comment by stonea** [comment]
> At a minimum the chapel build should probably check the CUDA version and fault if it's 12.x.

I've created an issue about doing this before building `chpl` itself:

https://github.com/chapel-lang/chapel/issues/22087

More generally, of course, we should get things to work with CUDA 12.x but it looks like we might have to wait around for clang to add support for it.

**Comment by e-kayrakli** [comment]
Thanks @stonea!

> As for what to do... That kinda depends on another question I have about the "best practices" for GPU installs. What I don't understand is how much of the chapel build relies on the CUDA that is specified at build time vs. how much (CUDA, arch, ...?) can be deferred and specified at application compile time. This isn't clear from the CPU technote. It would sure make life easier if I could build one GPU version of chapel and then mount it on the various GPU hosts and specify the CUDA/arch at application time. The alternative is that I have to build for each host/CUDA/arch version.

I agree that we can make the technote a bit clearer about that. But as you've observed, both CUDA version and the architecture matter only at application compilation time. We don't link CUDA during runtime build. And you can control GPU architecture with the `--gpu-arch` compiler flag as well as the environment variable. The CUDA path is only controlled by the environment variable, though. I don't think there's a deep reason why that is the case. Probably it is less common to want to change the path as opposed to the architecture?? But I don't know how much I believe that.

There are things we should improve both in terms of packaging/build scripts and documentation, but I think fundamentally you can use a Chapel build with different architectures and CUDA (minor?) versions.

**Comment by stonea** [comment]
I have a PR to add version requirements to the technote (re: CUDA and ROCM) here: https://github.com/chapel-lang/chapel/pull/22094

**Comment by stonea** [comment]
And I've merged a PR to actually validate the requirements as part of the build process here: https://github.com/chapel-lang/chapel/pull/22170

**Comment by bradcray** [comment]
Just to make sure I'm following the status of this issue:  My understanding is that Andy's PRs have put us in a better place, but we're keeping this issue open until we support CUDA 12.  Is that right?

**Comment by stonea** [comment]
Right, my PR improves things (gives a nicer error message) but doesn't solve the underlying issue.

We're thinking this is due to LLVM (for the NVPTX backend) not officially supporting a newer CUDA version than 10.1 (https://releases.llvm.org/12.0.0/docs/CompileCudaWithLLVM.html#prerequisites), although if you user a newer version it will warn about it and attempt to use it anyway. @hpcpony sees the "no template named 'texture'" error when using CUDA 12 on one machine but not when using CUDA 11.7 on another.

I'm tempted to say let's just wait for LLVM to update and see if the issue resolves itself. Let me do some Googling first though...

Some notable things I found:

* There's a relatively recent post on the LLVM discourse where someone is encountering the issue with clang 14 and 15 (but not clang 12). There's no replies yet about it: https://discourse.llvm.org/t/compiling-cuda-programs-with-clang-14-15/70013

* There's this post on the NVIDIA developer forums where someone is encountering this error from code they wrote that uses the texture template. In a reply a user says the code for this template is surrounded by an `#ifdef` for `CUDACC`. I think this macro gets set when compiling `.cu` files with `nvcc` (maybe we need to pass in `-DCUDACC` ourselves?)

* This Github issue for opencv mentions this error when building against CUDA 12: <https://github.com/opencv/opencv_contrib/issues/3390>.  Looking at some linked issues/PRs it mentions that CUDA texture objects were deprecated in favor of CUDA texture references in the 2012 (CUDA SDK 5.0 toolkit) and then removed in CUDA 12. I'm guessing this explains it. In our case the issue is coming from within a header for clang itself so I'm not sure there's anything we can do beyond wait.

Before we either close this or agree to sit on it, let me see if I can reproduce the issue on our end.  Assuming I can then I think it makes sense to either keep this issue open or open another one to remind us to check on it again after LLVM updates. I don't really have a preference between these two options though.

**Comment by bradcray** [comment]
Thanks for the additional details.  I'm not advocating for closing this one before its time—just wanted to make sure I understood the status and what it would take to close it.  Sounds like a lot of it is out of our hands (which is good to know).

**Comment by stonea** [comment]
Ok, I updated my desktop machine to use CUDA 12. I can repro it Chapel but that's not even necessary, just using `clang++` with `-xcuda` is enough:

```
$ clang++ --version
clang version 14.0.6
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin

$ cat foo.cpp
#include <stdio.h>
int main(int argc, char *argv[]) { printf("Hello world"); return 0; }

$ clang++ -xcuda foo.cpp -o foo
clang-14: warning: CUDA version is newer than the latest supported version 11.5 [-Wunknown-cuda-version]
In file included from <built-in>:1:
In file included from /usr/lib/clang/14.0.6/include/__clang_cuda_runtime_wrapper.h:365:
/usr/lib/clang/14.0.6/include/__clang_cuda_texture_intrinsics.h:696:13: error: no template named 'texture'
            texture<__DataT, __TexT, cudaReadModeNormalizedFloat> __handle,
            ^
/usr/lib/clang/14.0.6/include/__clang_cuda_texture_intrinsics.h:709:13: error: no template named 'texture'
            texture<__DataT, __TexT, cudaReadModeElementType> __handle,
            ^
2 errors generated when compiling for sm_35.
```

So yeah I think we can rest assured this is a more general LLVM/clang issue and not a issue with how we invoke it internally or anything like that.

**Comment by stonea** [comment]
> @stonea : Given that we now think this is an issue outside of our hands, should we close it with an ain't broke / won't fix label, or is there something more we could do to protect users from running into this unawares?

I keep hesitating on closing this because I want to make sure we verify that this actually is resolved once LLVM does officialy support CUDA 12.

I think there's a more general "what process should the GPU sub team have in place to deal with LLVM updates" we should probably have. Anyway, I'll bring up that discussion in some other venue and close this issue in the meantime.

**Comment by bradcray** [comment]
> I keep hesitating on closing this because I want to make sure we verify that this actually is resolved once LLVM does officialy support CUDA 12.

Sorry, I think I made you repeat yourself by not remembering/re-reading back far enough in the comment stream.  It would be fine with me to leave it open until the issue is resolved in LLVM+CUDA (and if we leave this comment towards the end of this stream, I probably won't keep asking about this each time I'm putting a user issues report together).

Copying a comment made on slack:  Is there a way we could file a future for this that would notify us when this issue was resolved (for our default/preferred version of LLVM, that is?).  That would help us proactively realize when the issue was resolved.

**Comment by e-kayrakli** [comment]
>  Is there a way we could file a future for this that would notify us when this issue was resolved (for our default/preferred version of LLVM, that is?)

Can't think of any myself. This would require CUDA 12 to be installed and used by the Chapel build. But if 12 wasn't supported, you can't build Chapel.

**Comment by bradcray** [comment]
I was imagining a skipif that would only run the test in the event that CUDA 12 was installed and used, but am almost certainly missing a more subtle dependence than that.
